src/odemis/acq/orsay_milling.py

```diff
@@ -0,0 +1,243 @@
+
+    def do_milling(self):
+        """
+        Function that handles milling action
+        """
+        # for now the ip address is hardcoded, it will be replaced with scanner.parent.host
+        rect = self.rect
+        iteration = self.iteration
+        duration = self.duration
+        probe_size = self.probe_size
+        overlap = self.overlap
+
+        server = _get_orsay_connection(self.scanner)
+        miller = server.datamodel.HybridPatternCreator
+        self.miller = miller
+
+        miller.MillingActivationState.Subscribe(self.on_milling_state)
+
+        fov = self.scanner.horizontalFoV.value
+
+        # the Y axes in odemis and orsay are inverted, that is why centerOfMassY is always inverted
+        # hardcoded 0.5's in the formulas below are the shift relative to the center of the FoV
+        if rect[2] == rect[0] and rect[1] == rect[3]:
+            # mill a point
+            obj_to_mill = Point()
+            obj_to_mill.primaryAxisX = 0
+            obj_to_mill.secondaryAxisLength = 0
+        else:
+            # scanner.shape gives the resolution like 1024x1024
+            # it is assumed that the scanner.shape is always square
+            obj_to_mill = Rectangle()
+            obj_to_mill.primaryAxisOverlap = overlap[0]
+            obj_to_mill.secondaryAxisOverlap = overlap[1]
+            obj_to_mill.primaryAxisY = 0  # primaryAxisY is always 0 to keep the  rectangle straight
+            if rect[0] == rect[2]:
+                logging.debug("Milling a vertical line. Orsay doesn't like rectangles with 0 "
+                              "height. Line height automatically set equal to the probe size to fit a single line.")
+                obj_to_mill.primaryAxisX = probe_size / 2
+                obj_to_mill.secondaryAxisLength = (rect[3] - rect[1]) / 2 * fov
+            elif rect[1] == rect[3]:
+                logging.debug("Milling a horizontal line. Orsay doesn't like rectangles with 0 "
+                              "width. Line width automatically set equal to the probe size to fit a single line.")
+                obj_to_mill.primaryAxisX = (rect[2] - rect[0]) / 2 * fov
+                obj_to_mill.secondaryAxisLength = probe_size / 2
+            else:
+                obj_to_mill.primaryAxisX = (rect[2] - rect[0]) / 2 * fov
+                obj_to_mill.secondaryAxisLength = (rect[3] - rect[1]) / 2 * fov
+        # common attribute value for point and rectangle, hence outside of if block
+        obj_to_mill.NumberOfPasses = iteration
+        obj_to_mill.totalScanTime = duration * iteration
+
+        obj_to_mill.centerOfMassX = ((rect[2] + rect[0]) / 2 - 0.5) * fov
+        obj_to_mill.centerOfMassY = (0.5 - (rect[3] + rect[1]) / 2) * fov
+
+        lay = Layer()
+        lay.probeSize = probe_size
+        lay.AddForm(0, obj_to_mill)
+
+        proc = Procedure()
+        proc.AddLayer(0, lay)
+        proc.name = f"Odemis rectangle milling {time.time()}"
```

Comment:
Please add a comment that the procedure name may not include `:` (or whatever character you found out is not allowed)